File: 0567-permutation-in-string/0567-permutation-in-string.py
from collections import Counter


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        s1_len, s2_len = len(s1), len(s2)
        if s1_len > s2_len:
            return False
        s1_counter = Counter(s1)
        s2_counter = Counter()
        for i in range(s1_len):
            s2_counter[s2[i]] += 1
        for i in range(s2_len - s1_len):
            if s1_counter == s2_counter:
                return True
            s2_counter[s2[i]] -= 1
            s2_counter[s2[i + s1_len]] += 1
        return s1_counter == s2_counter

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n1 = s1.size();
        int n2 = s2.size();

        if (n1 > n2) {
            return false;
        }

        int left = 0;
        int right = 0;
        array<int, 26> freq1{{}};
        array<int, 26> freq2{{}};

        while (right < n1) {
            freq1[s1[right] - 'a'] += 1;
            freq2[s2[right] - 'a'] += 1;
            right += 1;
        }
        while (right < n2) {
            if (freq1 == freq2) {
                return true;
            }
            freq2[s2[left] - 'a'] -= 1;
            freq2[s2[right] - 'a'] += 1;
            left += 1;
            right += 1;
        }
        return freq1 == freq2;
    }
};
